autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py

```diff
@@ -490,23 +509,43 @@ def process_train_data(self, df, labels):
         # print("df.shape",df.shape)
         if num_array_cols != df.shape[1]:
             raise ValueError("Error during one-hot encoding data processing for neural network. Number of columns in df array does not match feature_arraycol_map.")
-        
+
         # print(self.feature_arraycol_map)
         self.feature_type_map = self._get_feature_type_map() # OrderedDict of feature-name -> feature_type string (options: 'vector', 'embed', 'language')
         # print(self.feature_type_map)
         return TabularNNDataset(df, self.feature_arraycol_map, self.feature_type_map,
                                 batch_size=self.params['batch_size'], num_dataloading_workers=self.params['num_dataloading_workers'],
                                 problem_type=self.problem_type, labels=labels, is_test=False)
 
-    def setup_trainer(self):
-        """ Set up stuff needed for training: 
+    def setup_trainer(self, train_dataset):
+        """ Set up stuff needed for training:
             optimizer, loss, and summary writer (for mxboard).
-            Network must first be initialized before this. 
+            Network must first be initialized before this.
         """
-        optimizer_opts = {'learning_rate': self.params['learning_rate'],  
+        optimizer_opts = {'learning_rate': self.params['learning_rate'],
             'wd': self.params['weight_decay'], 'clip_gradient': self.params['clip_gradient']}
+        if 'lr_scheduler' in self.params and self.params['lr_scheduler'] is not None:
```

Comment:
I'm a bit concerned that these extra hyperparameters feel tacked on such that they don't belong to the PR. I'm ok with approving it with these changes present this time, but in future we should avoid having disjoint functionality being added in a single PR, especially functionality which we have no way of automatically verifying it won't crash upon use, yet users could technically specify in their HPO spaces.

Reply:
This functionality can be useful for inference-constrained / distillation-like settings where the user's goal is just deploy a single neural network model. While the LRscheduler is typically not as robust across many diverse datasets by default, it can be used to improve the performance of the NN on a particular dataset, given substantial user-tweaking/HPO to tailor it to this data.  

Is your concern regarding code coverage in the unit tests? I fully support expanding our  coverage.  I can remove it from this PR if you'd like and add it in a separate PR?  but even if I add testing coverage for this functionality, there are still tons of other autogluon-options not being covered in the unit tests.

Reply:
It is fine for now. Primarily I just think in future we should look towards more focused PR's. I believe these HPO settings should be their own PR.

Reply:
Added partial coverage in unit test: test_tabularHPO_bagstack.py
It's hard to fully cover all hyperparameter combinations within reasonable time-limit in unit tests.